### PR TITLE
Fix Product categories, Product Tags & Keyword filter not working in Products block

### LINKED ISSUE #8249: Products Block: Filtering by Category or Tag doesn't Work
## Describe the bug

Filtering products by category or tag in the Products block doesn't work.

## To reproduce

1. Insert the Products (Beta) block in a post or page
2. In the block settings, enter a product category or tag under Filters
3. Update the post or page

## Expected behavior

For the post or page to display products belonging to the specified category or tag

## What happened

The block does not respect the filter and displays all products on the live post or page.

## Ticket

5877934-zd

### WordPress (please complete the following information):

-   WordPress version: 6.1.1
-   WooCommerce version: 7.3.0

### PR DESCRIPTION
**Pull request by imanish003**
This PR fix [this issue](https://github.com/woocommerce/woocommerce-blocks/issues/8249), according to which following filters set from Inspector Controls for Products block aren't working:
- Product Categories
- Product tags
- Keyword

Fixes woocommerce/woocommerce-blocks#8249 

### Testing

1. Create a new post
2. Add Products block inside Editor
3. Add Product categories and/or Product Tags and/or Keyword to Filters(see screenshot)
image
4. Check the front-end. Only products with added categories, tags, or keywords are showing.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [X] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Fix Product categories, Product Tags & Keyword filter not working in Products block

### CODE REVIEW

File: src/BlockTypes/ProductQuery.php [hunk 1]
```diff
@@ -474,15 +474,29 @@ function( $acc, $attribute ) {
...
 			'price_filter'        => $this->get_filter_by_price_query(),
 			'attributes_filter'   => $this->get_filter_by_attributes_query(),
 			'stock_status_filter' => $this->get_filter_by_stock_status_query(),
 			'rating_filter'       => $this->get_filter_by_rating_query(),
 		);
+
+		// "Product Categories" & "Product Tags" could be provided using "Filters" ToolsPanel available in Inspector Controls.
+		if ( isset( $query['tax_query'] ) ) {
+			$result['tax_query'] = $this->get_filter_by_product_categories_or_tags_query( $query['tax_query'] );
+		}
```
Comment (imanish003) [inline src/BlockTypes/ProductQuery.php]: @sunyatasattva @gigitux Can you please take a look at this PR? I have a query regarding these lines of code.

As I am not sure what else could be there in `$query['tax_query']`, therefore I have created `get_filter_by_product_categories_or_tags_query` function to extract queries related to only Product categories & Product tags.

I am wondering if we could use `$query['tax_query']` as it is 🤔  Is there any possibility of extra items in `$query['tax_query']` which we won't need & therefore we need to extract only what we need((For example, only Product categories & tags))?

Comment (imanish003) [reply src/BlockTypes/ProductQuery.php]: Also, I can see that there is [get_global_query](https://github.com/woocommerce/woocommerce-blocks/blob/1954437c4616fedd27a2c18b41134b5ef1cfd0b4/src/BlockTypes/ProductQuery.php#L761) function. But code of this function only runs if `is_custom_inherit_global_query_implementation_enabled` is true. 

Only this function extracts `Keyword` which could be provided using "Filters" ToolsPanel available in Inspector Controls. Don't we need `Keyword` when is_custom_inherit_global_query_implementation_enabled is false?

**Review by sunyatasattva — CHANGES_REQUESTED** [review]
I'm getting some strange behaviour.

Sometimes when I input a category, I get this console error, and autocomplete doesn't work. Since the category input is a core one, it might be a bug from there, but worth mentioning:

Screenshot 2023-02-03 at 18 56 42

Otherwise, when I input a broad category such as “Clothing”, in the front-end I only get the results of that category and not its subcategories. E.g. “Clothing > Hoodies” do not appear. I believe the expected result of when I choose a broad category, would be to show all products from that category, including children. What do you think?

Keywords and tags seem to work correctly.

File: src/BlockTypes/ProductQuery.php (hunk 1 above)
Comment (sunyatasattva) [reply src/BlockTypes/ProductQuery.php]: >I am wondering if we could use $query['tax_query'] as it is 🤔 Is there any possibility of extra items in $query['tax_query'] which we won't need & therefore we need to extract only what we need((For example, only Product categories & tags))?

I don't recall exactly the reasoning behind it, as @gigitux has worked on that part of the code, but I believe it is intentional. Perhaps something to do with the interaction with filters blocks if we take `tax_query` as it is?

Or maybe it was a prophylactic measure to enable the `custom_inherit_global_query`, which we eventually want to switch to.

> Don't we need Keyword when is_custom_inherit_global_query_implementation_enabled is false?

Yes, we definitely do. It must have been an oversight we didn't catch, as we were focused on implementing the custom inherit, before the scope was changed.

Comment (gigitux) [reply src/BlockTypes/ProductQuery.php]: I don't remember very well, tbh 😭. 

In any case, I think that we should not use the `$query['tax_query'] as it is` to be ready when we will switch to the custom inherit query logic. 

We already have the logic[ to merge the queries](https://github.com/woocommerce/woocommerce-blocks/blob/4eda72955108774835946da7974b92aebf77c59b/src/BlockTypes/ProductQuery.php/#L196-L197).

Comment (imanish003) [reply src/BlockTypes/ProductQuery.php]: Hi @gigitux 👋 

> In any case, I think that we should not use the $query['tax_query'] as it is to be ready when we will switch to the custom inherit query logic.

Got it, then the changes I made in this PR looks good to me. In short, I created `get_filter_by_product_categories_or_tags_query` function to extract `product_tag` & `product_cat` from the `tax_query` as you can see [here](https://github.com/woocommerce/woocommerce-blocks/blob/1954437c4616fedd27a2c18b41134b5ef1cfd0b4/src/BlockTypes/ProductQuery.php#L832-L869). Also, I extracted the keyword [here](https://github.com/woocommerce/woocommerce-blocks/blob/1954437c4616fedd27a2c18b41134b5ef1cfd0b4/src/BlockTypes/ProductQuery.php#L786-L788). Please let me know if you notice any problem with this implementation. 

> We already have the logic[ to merge the queries](https://github.com/woocommerce/woocommerce-blocks/blob/4eda72955108774835946da7974b92aebf77c59b/src/BlockTypes/ProductQuery.php/#L196-L197).

Right, but while calling `merge_queries` function, we were not passing `tax_query` for product categories, product tags & keyword. I made the changes in this PR to pass these to `merge_queries` function. 

Please let me know if I misunderstood something & if you see any problem with changes I made in this PR.

Comment (gigitux) [reply src/BlockTypes/ProductQuery.php]: I would create a dedicated function for the filters applied on the editor side. The documentation about `get_queries_by_applied_filters` function says:

```
Return queries that are generated by query args.
```

**Comment by imanish003** [comment]
Hi @sunyatasattva, Thanks for the feedback 🙌🏻

> Sometimes when I input a category, I get this console error, and autocomplete doesn't work. Since the category input is a core one, it might be a bug from there, but worth mentioning:

I tried to reproduce it by adding & removing the categories, but I wasn't able to reproduce it. Can you please share the steps to reproduce it? IMO We can report the issue on GB repo once we know how to reproduce this issue. 

> when I input a broad category such as “Clothing”, in the front-end I only get the results of that category and not its subcategories. E.g. “Clothing > Hoodies” do not appear. I believe the expected result of when I choose a broad category, would be to show all products from that category, including children. What do you think?

Looks like this issue is there in `Query Loop` block. When a category is selected, it makes sense to show all the sub-category products. IMO we should report this issue on GB repo because when post type is either `product` or `post`, categories can be selected & sub-categories product/post doesn't show on frontend. What do you think?

Although, we can also fix it by fetching the sub-categories items. But, IMO it should be fixed in `Query Loop` block instead.

File: src/BlockTypes/ProductQuery.php (hunk 1 above)
Comment (imanish003) [reply src/BlockTypes/ProductQuery.php]: @gigitux That makes sense. There are two sections of filters in Editor i.e. `Advance Filters` & `Filters`  (see screenshot)
image

As per my understanding, we are creating queries for Advance filters in [get_queries_by_attributes](https://github.com/woocommerce/woocommerce-blocks/blob/1954437c4616fedd27a2c18b41134b5ef1cfd0b4/src/BlockTypes/ProductQuery.php#L508) (which isn't obvious from the name of function). 

I am thinking to do one of the following:
1. Maybe we can rename this function to `get_filter_queries` & return queries for `Advance Filters` & `Filters` from this function.
2. Rename this function to `get_advance_filter_queries` to make it more readable and clear. And, create new function called `get_filter_queries` which will return queries for product categories, product tags & keyword i.e. all the filters. Then pass `get_filter_queries()` as argument [here](https://github.com/woocommerce/woocommerce-blocks/blob/1954437c4616fedd27a2c18b41134b5ef1cfd0b4/src/BlockTypes/ProductQuery.php#L159-L165).

Which of these you think would be better approach?

Comment (gigitux) [reply src/BlockTypes/ProductQuery.php]: > As per my understanding, we are creating queries for Advance filters in [get_queries_by_attributes](https://github.com/woocommerce/woocommerce-blocks/blob/1954437c4616fedd27a2c18b41134b5ef1cfd0b4/src/BlockTypes/ProductQuery.php#L508) (which isn't obvious from the name of function).

That function generates queries by the custom attributes that the merchant sets via the editor. Furthermore, I think that there should not be a relationship between the UI and the name of the function of the backend.

So, I would create another function to create a valid query object starting from the $tax_query object.

Comment (imanish003) [reply src/BlockTypes/ProductQuery.php]: > That function generates queries by the custom attributes that the merchant sets via the editor.

Got it, maybe we can rename the function to `get_queries_by_custom_attributes` just to make it more readable & clear. What do you think? 🤔

> That function generates queries by the custom attributes that the merchant sets via the editor. Furthermore, I think that there should not be a relationship between the UI and the name of the function of the backend.
So, I would create another function to create a valid query object starting from the $tax_query object.

That makes sense. I will create a new function. Thanks for the feedback 🙌🏻

Comment (gigitux) [reply src/BlockTypes/ProductQuery.php]: > What do you think?

It makes sense! 👍 

> That makes sense. I will create a new function. Thanks for the feedback 🙌🏻

💪

Comment (imanish003) [reply src/BlockTypes/ProductQuery.php]: @gigitux I made the changes in [this commit](https://github.com/woocommerce/woocommerce-blocks/pull/8377/commits/db26bdb7d86434f367bf854e2d9b56141154a17b). Would you like to have a quick look?

**Comment by gigitux** [comment]
He, @imanish003! I noticed that this fix doesn't work with custom taxonomies. This how to reproduce the issue:
- Install `Custom Post Type UI` plugin
- Create a new taxonomy (when you create it, you must attach it to the product type).
- Add a new term to the taxonomy.
- Create a new post with the PQ.
- In the section filters, you can find a new input related to the new taxonomy that you added before. Now, filter by the term that you created earlier. You can see that the editor preview and frontend don't work.

A video that shows the issue. I should see only one product, the `cap`.

https://user-images.githubusercontent.com/4463174/219335925-a99c7b5d-8949-428b-807b-f723c134344d.mov

**Comment by imanish003** [comment]
Hey @gigitux, That's an interesting corner case. Thanks for catching it 🙌🏻

I was using a static array for taxonomies:
```php
$product_taxonomies = [ 'product_cat', 'product_tag' ];
```

In [e6bd862](https://github.com/woocommerce/woocommerce-blocks/pull/8377/commits/e6bd862f5c06ae57eace987445c353d0c4442186) commit, I have used `get_taxonomies` function to get all product taxonomies, which fix the issue.
```php
$product_taxonomies = get_taxonomies( array( 'object_type' => array( 'product' ) ), 'names' );
```

Can you please have a look again?

File: src/BlockTypes/ProductQuery.php [hunk 2]
```diff
@@ -220,8 +222,8 @@ function( $acc, $query ) {
 		 * duplicated items.
 		 */
 		if (
-			! empty( $merged_query['post__in'] ) &&
-			count( $merged_query['post__in'] ) > count( array_unique( $merged_query['post__in'] ) )
+		! empty( $merged_query['post__in'] ) &&
```
Comment (gigitux) [inline src/BlockTypes/ProductQuery.php]: Not sure, but I guess that we should revert this formatting change.

File: src/BlockTypes/ProductQuery.php [hunk 3]
```diff
@@ -337,8 +339,8 @@ private function get_stock_status_query( $stock_statii ) {
 		 * meta query for stock status.
 		 */
 		if (
-			count( $stock_statii ) === count( $stock_status_options ) &&
```
Comment (gigitux) [inline src/BlockTypes/ProductQuery.php]: Not sure, but I guess that we should revert this formatting change.

File: src/BlockTypes/ProductQuery.php [hunk 4]
```diff
@@ -619,15 +621,15 @@ function( $stock_status ) {
...
-				array(
-					'key'      => '_stock_status',
-					'value'    => $filtered_stock_status_values,
-					'operator' => 'IN',
+		// phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_query
+		'meta_query' => array(
+			array(
+				'key'      => '_stock_status',
+				'value'    => $filtered_stock_status_values,
+				'operator' => 'IN',
```
Comment (gigitux) [inline src/BlockTypes/ProductQuery.php]: Not sure, but I guess that we should revert this formatting change.

File: src/BlockTypes/ProductQuery.php [hunk 5]
```diff
@@ -814,4 +816,61 @@ function( $rating ) use ( $product_visibility_terms ) {
...
+
+		return ! empty( $result ) ? [ 'tax_query' => $result ] : [];
+	}
+
+	/**
+	 * Returns the keyword filter from the given query.
+	 *
+	 * @param WP_Query $query The query to extract the keyword filter from.
+	 * @return array The keyword filter, or an empty array if none is found.
+	 * @throws InvalidArgumentException If $query is not an array.
```
Comment (gigitux) [inline src/BlockTypes/ProductQuery.php]: I thank that we should do this check, in this way, we will not throw any errors. Now we don't catch the error.

File: src/BlockTypes/ProductQuery.php [hunk 6]
```diff
@@ -160,8 +160,10 @@ public function build_query( $query ) {
 			$common_query_values,
 			$this->get_global_query( $parsed_block ),
 			$this->get_custom_orderby_query( $query['orderby'] ),
-			$this->get_queries_by_attributes( $parsed_block ),
-			$this->get_queries_by_applied_filters()
+			$this->get_queries_by_custom_attributes( $parsed_block ),
+			$this->get_queries_by_applied_filters(),
+			$this->get_filter_by_taxonomies_query( $query['tax_query'] ),
```
Comment (gigitux) [inline src/BlockTypes/ProductQuery.php]:
```suggestion
			$this->get_filter_by_taxonomies_query( isset($query['tax_query'] ? $query['tax_query']: array()),
```

Now, if the Products block doesn't have any filter defined, `tax_query` doesn't exist, and WP returns an error.

**Review by gigitux — CHANGES_REQUESTED** [review]
Thanks for fixing the issue! 💪  I added some comments, after addressing them I think that we can merge the PR!

**Comment by imanish003** [comment]
Hey @gigitux, I truly appreciate your attention to detail 🚀 I have made the necessary revisions based on your recommendations. Can you please have a look again?

**Review by gigitux — APPROVED** [review]
LGTM! Thanks for addressing the feedback! :shipit: 

Also, I suggest to write E2E tests that check those use-cases. It could make sense do it in a follow-up PR 👍

**Comment by imanish003** [comment]
@gigitux That makes sense. I will create a follow-up PR. I have created [this task](https://github.com/woocommerce/woocommerce/issues/42473) for E2E tests. 🙌🏻

**Pull request merged**